File: apps/ios/SignalASI/AgentTaskIdentity.swift
import CryptoKit
import Foundation

struct AgentTaskIdentity: Codable, Equatable {
  var clientRouteId: String
  var conversationId: String
  var taskId: String
  var turnId: String

  var isComplete: Bool {
    !isBlank(clientRouteId) &&
      !isBlank(conversationId) &&
      !isBlank(taskId) &&
      !isBlank(turnId)
  }

  init(
    clientRouteId: String,
    conversationId: String,
    taskId: String,
    turnId: String
  ) {
    self.clientRouteId = clientRouteId
    self.conversationId = conversationId
    self.taskId = taskId
    self.turnId = turnId
  }

  enum CodingKeys: String, CodingKey {
    case clientRouteId = "client_route_id"
    case conversationId = "conversation_id"
    case taskId = "task_id"
    case turnId = "turn_id"
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    clientRouteId = try container.decodeIfPresent(String.self, forKey: .clientRouteId) ?? ""
    conversationId = try container.decodeIfPresent(String.self, forKey: .conversationId) ?? ""
    taskId = try container.decodeIfPresent(String.self, forKey: .taskId) ?? ""
    turnId = try container.decodeIfPresent(String.self, forKey: .turnId) ?? ""
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encode(clientRouteId, forKey: .clientRouteId)
    try container.encode(conversationId, forKey: .conversationId)
    try container.encode(taskId, forKey: .taskId)
    try container.encode(turnId, forKey: .turnId)
  }

  private func isBlank(_ value: String) -> Bool {
    value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
  }
}

enum AgentTaskIdentityPolicy {
  static func conversationId(contactId: String, requested: String) -> String {
    trim(requested).isEmpty ? "contact:\(trim(contactId))" : trim(requested)
  }

  static func turnId(
    sourceMessageId: Int64?,
    requested: String,
    fallbackUUID: () -> UUID = { UUID() }
  ) -> String {
    let explicit = trim(requested)
    if !explicit.isEmpty {
      return explicit
    }
    if let sourceMessageId, sourceMessageId > 0 {
      return "message:\(sourceMessageId)"
    }
    return fallbackUUID().uuidString.lowercased()
  }

  static func taskId(
    ownerId: String,
    contactId: String,
    sourceMessageId: Int64?,
    conversationId: String,
    turnId: String,
    requested: String = ""
  ) -> String {
    let explicit = trim(requested)
    if !explicit.isEmpty {
      return explicit
    }
    let seed = [
      trim(ownerId),
      trim(contactId),
      sourceMessageId.map { String($0) } ?? "",
      trim(conversationId),
      trim(turnId)
    ].joined(separator: "\u{001f}")
    return nameBasedUUID(seed).uuidString.lowercased()
  }

  static func taskId(
    ownerId: String,
    contactId: String,
    sourceMessageId: String,
    conversationId: String,
    turnId: String,
    requested: String = ""
  ) -> String {
    let explicit = trim(requested)
    if !explicit.isEmpty {
      return explicit
    }
    let seed = [
      trim(ownerId),
      trim(contactId),
      trim(sourceMessageId),
      trim(conversationId),
      trim(turnId)
    ].joined(separator: "\u{001f}")
    return nameBasedUUID(seed).uuidString.lowercased()
  }

  static func matchesDesktopResponse(
    expected: [String: String],
    conversationId: String,
    taskId: String,
    turnId: String
  ) -> Bool {
    guard expected["resource_location"] == "desktop" else { return true }
    let expectedConversationId = expected["conversation_id"] ?? ""
    let expectedTaskId = expected["remote_task_id"] ?? ""
    let expectedTurnId = expected["turn_id"] ?? ""
    return !isBlank(expectedConversationId) &&
      !isBlank(expectedTaskId) &&
      !isBlank(expectedTurnId) &&
      conversationId == expectedConversationId &&
      taskId == expectedTaskId &&
      turnId == expectedTurnId
  }

  static func routesToMainAgent(
    superseded: Bool,
    hasRuntime: Bool,
    resolvedConversationId: String
  ) -> Bool {
    superseded || hasRuntime || !isBlank(resolvedConversationId)
  }

  static func matchesResponseIdentity(
    expectedConversationId: String,
    expectedTurnId: String,
    expectedTaskId: String,
    actualConversationId: String,
    actualTurnId: String,
    actualTaskId: String
  ) -> Bool {
    let expectedConversationId = trim(expectedConversationId)
    let expectedTurnId = trim(expectedTurnId)
    let expectedTaskId = trim(expectedTaskId)
    let actualConversationId = trim(actualConversationId)
    let actualTurnId = trim(actualTurnId)
    let actualTaskId = trim(actualTaskId)
    let expectedHasTurnIdentity = !expectedConversationId.isEmpty || !expectedTurnId.isEmpty
    let actualHasTurnIdentity = !actualConversationId.isEmpty || !actualTurnId.isEmpty
    if !expectedHasTurnIdentity && !actualHasTurnIdentity { return true }
    if !expectedHasTurnIdentity || !actualHasTurnIdentity { return false }
    guard expectedConversationId == actualConversationId,
          expectedTurnId == actualTurnId else {
      return false
    }
    return expectedTaskId.isEmpty || actualTaskId.isEmpty || expectedTaskId == actualTaskId
  }

  private static func nameBasedUUID(_ value: String) -> UUID {
    var bytes = Array(Insecure.MD5.hash(data: Data(value.utf8)))
    bytes[6] = (bytes[6] & 0x0f) | 0x30
    bytes[8] = (bytes[8] & 0x3f) | 0x80
    return UUID(uuid: (
      bytes[0], bytes[1], bytes[2], bytes[3],
      bytes[4], bytes[5],
      bytes[6], bytes[7],
      bytes[8], bytes[9],
      bytes[10], bytes[11], bytes[12], bytes[13], bytes[14], bytes[15]
    ))
  }

  private static func trim(_ value: String) -> String {
    value.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private static func isBlank(_ value: String) -> Bool {
    trim(value).isEmpty
  }
}

final class AgentTaskIdentityStore {
  private let defaults: UserDefaults
  private let storageKeyPrefix: String
  private let encoder = JSONEncoder()
  private let decoder = JSONDecoder()
  private let lock = NSRecursiveLock()

  init(
    defaults: UserDefaults = .standard,
    storageKeyPrefix: String = "signalasi_agent_task_identities"
  ) {
    self.defaults = defaults
    self.storageKeyPrefix = storageKeyPrefix
  }

  func register(
    contactId: String,
    sourceMessageId: Int64,
    identity: AgentTaskIdentity
  ) {
    guard sourceMessageId > 0 else { return }
    register(
      contactId: contactId,
      sourceMessageId: String(sourceMessageId),
      identity: identity
    )
  }

  func register(
    contactId: String,
    sourceMessageId: String,
    identity: AgentTaskIdentity
  ) {
    guard let contactId = clean(contactId),
          let sourceMessageId = clean(sourceMessageId),
          identity.isComplete,
          let encoded = try? encoder.encode(identity) else {
      return
    }
    lock.lock()
    defer { lock.unlock() }
    defaults.set(
      String(decoding: encoded, as: UTF8.self),
      forKey: storageKey(contactId, sourceMessageId)
    )
  }

  func matches(payload: [String: Any]) -> Bool {
    matchesStored(payload: payload, requireRegistered: false)
  }

  func matchesRegistered(payload: [String: Any]) -> Bool {
    matchesStored(payload: payload, requireRegistered: true)
  }

  func identity(contactId: String, sourceMessageId: String) -> AgentTaskIdentity? {
    guard let contactId = clean(contactId),
          let sourceMessageId = clean(sourceMessageId) else {
      return nil
    }
    lock.lock()
    defer { lock.unlock() }
    guard let encoded = defaults.string(forKey: storageKey(contactId, sourceMessageId)) else {
      return nil
    }
    return decodeIdentity(encoded)
  }

  private func matchesStored(payload: [String: Any], requireRegistered: Bool) -> Bool {
    guard let contactId = clean(payload.string("contact_id")),
          let sourceMessageId = clean(sourceMessageIdentity(from: payload)) else {
      return false
    }
    lock.lock()
    defer { lock.unlock() }
    let encoded = defaults.string(forKey: storageKey(contactId, sourceMessageId))
    guard let encoded else {
      return !requireRegistered
    }
    guard let expected = decodeIdentity(encoded) else {
      return false
    }
    return payload.string("client_route_id") == expected.clientRouteId &&
      payload.string("conversation_id") == expected.conversationId &&
      payload.string("task_id") == expected.taskId &&
      payload.string("turn_id") == expected.turnId
  }

  private func sourceMessageIdentity(from payload: [String: Any]) -> String {
    let source = payload.string("source_message_id")
      .trimmingCharacters(in: .whitespacesAndNewlines)
    if !source.isEmpty {
      return source
    }
    return payload.string("client_message_id")
      .trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private func decodeIdentity(_ encoded: String) -> AgentTaskIdentity? {
    guard let data = encoded.data(using: .utf8) else { return nil }
    return try? decoder.decode(AgentTaskIdentity.self, from: data)
  }

  private func storageKey(_ contactId: String, _ sourceMessageId: String) -> String {
    let raw = "\(contactId)\u{001f}\(sourceMessageId)"
    let digest = SHA256.hash(data: Data(raw.utf8))
      .map { String(format: "%02x", $0) }
      .joined()
    return "\(storageKeyPrefix).\(digest)"
  }

  private func clean(_ value: String) -> String? {
    let clean = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return clean.isEmpty ? nil : clean
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentTaskIdentity.kt
package com.galaxyssi.chat

import android.content.Context
import java.nio.charset.StandardCharsets
import java.util.UUID
import org.json.JSONObject

internal data class AgentTaskIdentity(
    val clientRouteId: String,
    val conversationId: String,
    val taskId: String,
    val turnId: String
) {
    val isComplete: Boolean
        get() = clientRouteId.isNotBlank() &&
            conversationId.isNotBlank() &&
            taskId.isNotBlank() &&
            turnId.isNotBlank()
}

internal data class AgentConnectorResponseIdentity(
    val conversationId: String,
    val taskId: String,
    val turnId: String
)

internal object AgentTaskIdentityPolicy {
    fun conversationId(contactId: String, requested: String): String =
        requested.trim().ifBlank { "contact:${contactId.trim()}" }

    fun turnId(sourceMessageId: Long?, requested: String): String =
        requested.trim().ifBlank {
            sourceMessageId
                ?.takeIf { it > 0L }
                ?.let { "message:$it" }
                ?: UUID.randomUUID().toString()
        }

    fun taskId(
        ownerId: String,
        contactId: String,
        sourceMessageId: Long?,
        conversationId: String,
        turnId: String,
        requested: String = ""
    ): String {
        val explicit = requested.trim()
        if (explicit.isNotBlank()) return explicit
        val seed = listOf(
            ownerId.trim(),
            contactId.trim(),
            sourceMessageId?.toString().orEmpty(),
            conversationId.trim(),
            turnId.trim()
        ).joinToString("\u001f")
        return UUID.nameUUIDFromBytes(seed.toByteArray(StandardCharsets.UTF_8)).toString()
    }

    fun matchesDesktopResponse(
        expected: Map<String, String>,
        conversationId: String,
        taskId: String,
        turnId: String
    ): Boolean {
        if (expected["resource_location"] != "desktop") return true
        val expectedConversationId = expected["conversation_id"].orEmpty()
        val expectedTaskId = expected["remote_task_id"].orEmpty()
        val expectedTurnId = expected["turn_id"].orEmpty()
        return expectedConversationId.isNotBlank() &&
            expectedTaskId.isNotBlank() &&
            expectedTurnId.isNotBlank() &&
            conversationId == expectedConversationId &&
            taskId == expectedTaskId &&
            turnId == expectedTurnId
    }

    fun canonicalConnectorResponseIdentity(
        pendingDelivery: AgentPendingDelivery?,
        conversationId: String,
        taskId: String,
        turnId: String
    ): AgentConnectorResponseIdentity = AgentConnectorResponseIdentity(
        conversationId = pendingDelivery?.conversationId.orEmpty().ifBlank {
            conversationId.trim()
        },
        taskId = pendingDelivery?.taskId.orEmpty().ifBlank {
            taskId.trim()
        },
        turnId = pendingDelivery?.turnId.orEmpty().ifBlank {
            turnId.trim()
        }
    )

    fun routesToMainAgent(
        superseded: Boolean,
        hasRuntime: Boolean,
        resolvedConversationId: String
    ): Boolean = superseded || hasRuntime || resolvedConversationId.isNotBlank()
}

internal object AgentTaskIdentityStore {
    private const val PREFS_NAME = "galaxyssi_agent_task_identities"

    fun register(
        context: Context,
        contactId: String,
        sourceMessageId: Long,
        identity: AgentTaskIdentity
    ) {
        if (contactId.isBlank() || sourceMessageId <= 0L || !identity.isComplete) return
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putString(
                key(contactId, sourceMessageId),
                JSONObject()
                    .put("client_route_id", identity.clientRouteId)
                    .put("conversation_id", identity.conversationId)
                    .put("task_id", identity.taskId)
                    .put("turn_id", identity.turnId)
                    .toString()
            )
            .apply()
    }

    fun matches(context: Context, payload: JSONObject): Boolean {
        return matchesStored(context, payload, requireRegistered = false)
    }

    fun matchesRegistered(context: Context, payload: JSONObject): Boolean {
        return matchesStored(context, payload, requireRegistered = true)
    }

    private fun matchesStored(
        context: Context,
        payload: JSONObject,
        requireRegistered: Boolean
    ): Boolean {
        val contactId = payload.optString("contact_id")
        val sourceMessageId = payload.optString("source_message_id").toLongOrNull()
            ?: payload.optLong("source_message_id", 0L)
        if (contactId.isBlank() || sourceMessageId <= 0L) return false
        val encoded = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getString(key(contactId, sourceMessageId), null)
            ?: return !requireRegistered
        val expected = runCatching { JSONObject(encoded) }.getOrNull() ?: return false
        return payload.optString("client_route_id") == expected.optString("client_route_id") &&
            payload.optString("conversation_id") == expected.optString("conversation_id") &&
            payload.optString("task_id") == expected.optString("task_id") &&
            payload.optString("turn_id") == expected.optString("turn_id")
    }

    private fun key(contactId: String, sourceMessageId: Long): String =
        "${contactId.trim()}\u001f$sourceMessageId"
}
